### PR TITLE
Nibble storage class

### PR DESCRIPTION
**Pull request by betatim**
This introduces a storage class that uses 4bit per counter.

The storage class is `NibbleStorage` and it is used in `SmallCounttable`.

Todo:
- [x] save/load
- [ ] multi threading?!
- [x] big count and/or overflow?
- [x] `SmallCountgraph` ?
- [x] check conversion of size arguments

Checklist:
- [x] Is it mergeable?
- [x] `make test` Did it pass the tests?
- [x] `make clean diff-cover` If it introduces new functionality in
  `scripts/` is it tested?
- [x] `make format diff_pylint_report cppcheck doc pydocstyle` Is it well
  formatted?
- [x] Did it change the command-line interface? Only backwards-compatible
  additions are allowed without a major version increment. Changing file
  formats also requires a major version number increment.
- [x] For substantial changes or changes to the command-line interface, is it
  documented in `CHANGELOG.md`? See [keepachangelog](http://keepachangelog.com/)
  for more details.
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?
- [x] Do the changes respect streaming IO? (Are they
  tested for streaming IO?)

### CODE REVIEW

**Comment by betatim** [comment]
Addresses part of #1035

**Comment by ctb** [comment]
Looks good on a first skim -- is this something that should be merged into #1511, do you think, or is it best => master? If into #1511 it would be able to take advantage of k > 32.

**Comment by betatim** [comment]
I'm easy either way. Merge this into master and then rebase #1511 or merge this into #1511 and then later into master. I'd pick the option where it gets used sooner.

**Comment by betatim** [comment]
Now the counter doesn't overflow anymore but stops counting up at 15. This seems much more sensible than overflowing. It means you can run
```
$ trim-low-abund.py -k 32 -M 1e6 -C 3 -V -Z 15 ../zaok/data/mixed-reads.fa.gz -o temp/mixed-trimV-2.fa
```
using the same amount of memory but with bigger tables (-> smaller fp) or less memory (at constant fp):
```
$ trim-low-abund.py -k 32 -M 1e6 -C 3 -V -Z 15 ../zaok/data/mixed-reads.fa.gz -o temp/mixed-trimV-2.fa
```

File: lib/storage.hh [hunk 1]
```diff
@@ -237,6 +237,7 @@ public:
 class NibbleStorage : public Storage
 {
 protected:
+    // table size is measured in number of entries in the table, not in bytes
```
Comment (standage) [inline lib/storage.hh]: I was just gonna ask about this. Thanks for clarifying!

**Comment by betatim** [comment]
@standage do you have an opinion on where to merge this to? I assume you'd be one of the keen first users.

**Comment by standage** [comment]
@betatim I'd say directly to master. Definitely related to #1511, but orthogonal enough to warrant separate consideration.

**Comment by betatim** [comment]
@standage do you have time for a review?

There is one outstanding question: how to do this when the code is running on multiple threads. The simple answer is locks. So that is still "todo" but otherwise I think ready for a second pair of eyes.

File: tests/test_countgraph.py [hunk 2]
```diff
@@ -218,6 +229,18 @@ def test_get_raw_tables_view():
...
+def test_get_raw_tables_view_smallcountgraph():
+    ht = khmer.SmallCountgraph(20, 1e5, 4)
+    tables = ht.get_raw_tables()
+    for tab in tables:
+        assert sum(tab.tolist()) == 0
+    ht.consume('AAAATTTTCCCCGGGGAAAA')
+    # the actual count is 1 but stored in the first 4bits of a Byte
+    # and so becomes 16
+    for tab in tables:
+        assert sum(tab.tolist()) == int('00010000', 2)
```
Comment (standage) [inline tests/test_countgraph.py]: 👍

File: tests/test_countgraph.py [hunk 3]
```diff
@@ -634,6 +657,27 @@ def do_test(ctfile):
...
+
+        loaded = khmer.load_countgraph(savepath, small=True)
+
+        orig_count = orig.n_occupied()
+        loaded_count = loaded.n_occupied()
+        assert orig_count == 3881, orig_count
+        assert loaded_count == orig_count, loaded_count
+
+    for ctfile in ['temp.ct', 'temp.ct.gz']:
+        do_test(ctfile)
```
Comment (standage) [inline tests/test_countgraph.py]: @ctb Isn't a fan of loops in tests, although I must admit I'm not sure why. Perhaps it's just a matter of keeping them simple? I'm ok with this, but I'll let him comment on it.

File: tests/test_nibblestorage.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    sct = SmallCounttable(4, 1e6, 4)
+    a = "AAAA"
+    b = "AAAT"
+
+    # try to overflow our 4bit counter
+    for n in range(17):
+        sct.add(a)
+
+    assert sct.get(a) == 15
+    assert sct.get(b) == 0
```
Comment (standage) [inline tests/test_nibblestorage.py]: 👍

**Comment by standage** [comment]
Shall I hold off with merging until multithreading support is there?

**Comment by ctb** [comment]
On Wed, Dec 07, 2016 at 11:09:23AM -0800, Daniel Standage wrote:
> Shall I hold off with merging until multithreading support is there?

let me take a look.

**Comment by ctb** [comment]
On Wed, Dec 07, 2016 at 11:07:51AM -0800, Daniel Standage wrote:
> +    for ctfile in ['temp.ct', 'temp.ct.gz']:
> +        do_test(ctfile)
> 
> @ctb Isn't a fan of loops in tests, although I must admit I'm not sure why. Perhaps it's just a matter of keeping them simple? I'm ok with this, but I'll let him comment on it.

Exactly - we shouldn't have to debug our test code, ya know?  Loops and
conditionals make programming difficult :)

Plus, when a test fails, the for loop makes figuring out which part of the
test triggered the problem tricky.

**Comment by betatim** [comment]
Tests got parametrised to remove the loops.

**Comment by ctb** [comment]
It seems like the last remaining concern here is thread safety - @betatim, thoughts?

File: khmer/khmer_args.py [hunk 5]
```diff
@@ -479,13 +487,16 @@ def add_loadgraph_args(parser):
 
 def calculate_graphsize(args, graphtype, multiplier=1.0):
     """Transform the table parameters into a size."""
-    if graphtype not in ('countgraph', 'nodegraph'):
+    if graphtype not in ('countgraph', 'smallcountgraph', 'nodegraph'):
```
Comment (ctb) [inline khmer/khmer_args.py]: *net

File: lib/storage.hh [hunk 6]
```diff
@@ -221,6 +221,175 @@ public:
...
+    // which you then need to select the correct nibble from
+    uint64_t _table_index(const HashIntoType k, const uint64_t tablesize) const
+    {
+        const uint64_t bins = tablesize / 2 + 1;
+        return (k / 2) % bins;
+    }
+    // Compute which half of the byte to use for this hash value
+    uint8_t _mask(const HashIntoType k) const
+    {
+        return k%2 ? 15 : 240;
```
Comment (ctb) [inline lib/storage.hh]: This may be premature optimization, but instead of k%2 what about just looking at the first bit?

**Comment by betatim** [comment]
Yes thread safety is missing. Thinking we can do what the other storages are doing. Somehow got distracted by #1248 

Let's wrap this up before 🎄

**Comment by betatim** [comment]
The latest commit is semi related to the actual content of the PR.

While trying to track down the source of #1248 and thinking about multithreading for `NibbleStorage` I ended up with 61a07a1 as the only way to fix #1248. In which case we don't need anything threadsafe in `NibbleStorage`. Can someone help me understand why locking neither, just `get_next_read` or just `check_and_process_read` don't work?

I used `for n in {1..20}; do time python scripts/abundance-dist-single.py --threads 2 -s -x 1e7 -N 2 -k 17 -z tests/test-data/test-large.fa /tmp/test.dist 2>&1 | grep Total; done` to establish whether or not something was thread safe.

**Comment by standage** [comment]
@betatim Sorry, only experience I have with parallel processing is OpenMP. I did some copy-n-paste today with #1571 but I can't say I have more than a superficial understanding of thread-safe implementation.

**Comment by betatim** [comment]
Solution to the difference in n_unique_kmers for ByteStorage and NibbleStorage is in `0b2fa13`. Basically by splitting a given byte nible storage always had an even number of buckets. current double modulus isn't pretty or fast but it is correct (famous last words).

**Comment by betatim** [comment]
I've ticked the multi-threading box. You can run `python scripts/abundance-dist-single.py --small-count --threads 5 -s -x 5e7 -N 2 -k 29 -z tests/test-data/test-large.fa /tmp/test.dist` (or without `--small-count`) and get a consistent number of unique k-mers. If you remove the mutex introduced in c58293e neither `Counttable` nor `SmallCounttable` manage to give a consistent answer over multiple runs.

**Comment by ctb** [comment]
@betatim I don't understand why we need a mutex guard around the read retrieval code...

**Comment by betatim** [comment]
I don't understand it either but empirically you need it otherwise the number of unique kmers you find changes :-/

**Comment by ctb** [comment]
that is a very unsatisfying answer :)

**Comment by ctb** [comment]
Seems to me like this could be merged. Update from master & merge?

**Comment by standage** [comment]
Yeppers.

**Comment by betatim** [comment]
:+1: 

(I can't convince myself either way whether we need the mutex or not. Basically, can we observe the mistakes from race conditions or not? Maybe something for a separate issue)

**Comment by ctb** [comment]
yes, bump to issue with brief summary, pls!

**Pull request merged**